tests/specs/lemmas.k

```diff
@@ -241,6 +241,12 @@ module LEMMAS-HASKELL [symbolic, kore]
     imports EVM
     imports EDSL
 
+    rule #lookup ( M:Map [ K1 <- V1 ] , K2 ) => V1                 requires K1 ==K  K2
+    rule #lookup ( M:Map [ K1 <- _  ] , K2 ) => #lookup ( M , K2 ) requires K1 =/=K K2
+
+    rule ( M:Map [ K1 <- _  ] ) [ K2 <- V2 ] =>   M [ K1 <- V2 ]                requires K1 ==K  K2
+    rule ( M:Map [ K1 <- V1 ] ) [ K2 <- V2 ] => ( M [ K2 <- V2 ] ) [ K1 <- V1 ] requires K1 =/=K K2
```

Comment:
This (the last rule) could be problematic for the debugging purpose, as it could swap the order of updates (but in a sound way).  To preserve the order, we need to check if `K1 in keys(M)`, but the current haskell backend seems to have some issues regarding the reasoning about `in_keys`.

Reply:
1. For first 2 rules I have this version in erc20-verifier, which is probably faster but it doesn't work, because there are bugs in Haskell:
```
  rule #lookup( M [KEY1 <- VAL:Int], KEY2 )
       => #if KEY1 ==K KEY2 #then VAL #else #lookup( M, KEY2 ) #fi                    [simplification]
```
So if you need this change now, then (a) add `V1:Int` (b) add [simplification] everywhere
. If not, then maybe it's better to postpone until Haskell gets fixed.

2. For 4th rule, looks to me like it'll cause infinite loop. Waht exactly are you verifying.

Reply:
The last rule should be removed. I use these rules for ordering updates and pushing single-value updates inside of range updates on the MCD proofs:

```
    // Memory assignment sorting/normalization
    rule MAP [ N := .ByteArray ] => MAP
    rule MAP [ N := W : WS     ] => MAP [ N <- W ] [ N +Int 1 := WS ]

    rule MAP:Memory [ N <- V           ] [ N' <- V' ] => MAP [ N' <- V' ] [ N <- V ] requires N'                          <Int N
    rule MAP:Memory [ N := V:ByteArray ] [ N' <- V' ] => MAP [ N' <- V' ] [ N := V ] requires N'                          <Int N orBool N +Int #sizeByteArray(V) <=Int N'
    rule MAP:Memory [ N := V:ByteArray ] [ N' := V' ] => MAP [ N' := V' ] [ N := V ] requires N' +Int #sizeByteArray(V') <=Int N
    rule MAP:Memory [ N <- V           ] [ N' := V' ] => MAP [ N' := V' ]            requires N' <=Int N andBool N <Int N' +Int #sizeByteArray(V')
```

Reply:
These rules I ahve found tend to minimize the size of the map terms that show up.

Reply:
OK. The last rule is problematic. Thanks @denis-bogdanas. 

@denis-bogdanas Regarding your simpler lookup rule, could you elaborate more on it?  Why does it work?  Also, what's the current issue of the haskell backend?

Reply:
@ehildenb Your rules are for the integer-key maps, right?  My intention was to have rules for arbitrary maps.  Originally I would like to add the following, but wasn't sure if it works for the haskell backend:
```
rule ( M:Map [ K1 <- V1 ] ) [ K2 <- V2 ] => ( M [ K2 <- V2 ] ) [ K1 <- V1 ] requires K1 =/=K K2 andBool K1 in_keys(M)
```